src/bridge/cockpitdbusmachines.c

```diff
@@ -0,0 +1,416 @@
+  json_generator_set_root (json_gen, cur_config);
+  json_generator_set_pretty (json_gen, TRUE); /* bikeshed zone */
+  res = json_generator_to_file (json_gen, path, error);
+  g_free (path);
+  g_object_unref (json_gen);
+  json_node_free (cur_config);
+  return res;
+}
+
+static GVariant *
+machines_get_property (GDBusConnection *connection,
+                       const gchar *sender,
+                       const gchar *object_path,
+                       const gchar *interface_name,
+                       const gchar *property_name,
+                       GError **error,
+                       gpointer user_data)
+{
+  g_return_val_if_fail (property_name != NULL, NULL);
+
+  if (g_str_equal (property_name, "Machines"))
+    return build_machines ();
+  else
+    g_return_val_if_reached (NULL);
+}
+
+static void
+machines_method_call (GDBusConnection *connection,
+                      const gchar *sender,
+                      const gchar *object_path,
+                      const gchar *interface_name,
+                      const gchar *method_name,
+                      GVariant *parameters,
+                      GDBusMethodInvocation *invocation,
+                      gpointer user_data)
+{
+  if (g_str_equal (method_name, "Update"))
+    {
+      const gchar *filename, *hostname;
+      GVariant * info_v;
+      JsonNode *info_json = NULL;
+      GError *error = NULL;
+
+      g_variant_get (parameters, "(&s&s@a{sv})", &filename, &hostname, &info_v);
+      info_json = json_gvariant_serialize (info_v);
+      /* g_debug ("Updating %s for machine %s: %s", filename, hostname, g_variant_print (info_v, TRUE)); */
+      g_variant_unref (info_v);
+
+      if (update_machine (filename, hostname, info_json, &error))
+        g_dbus_method_invocation_return_value (invocation, NULL);
+      else
+        g_dbus_method_invocation_take_error (invocation, error);
+      json_node_free (info_json);
+    }
+  else
+    g_return_if_reached ();
+}
+
+static void
+on_machines_changed (GFileMonitor     *monitor,
+                     GFile            *file,
+                     GFile            *other_file,
+                     GFileMonitorEvent event_type,
+                     gpointer          user_data)
+{
+  gchar *path;
+  GDBusConnection *connection = user_data;
+  GVariant *machines, *signal_value;
+  GVariantBuilder builder;
+  GError *error = NULL;
+
+  /* ignore uninteresting events; note that DELETED does not get a followup CHANGES_DONE_HINT */
+  if (event_type != G_FILE_MONITOR_EVENT_CHANGES_DONE_HINT && event_type != G_FILE_MONITOR_EVENT_DELETED)
+    return;
+
+  path = g_file_get_path (file);
+  if (g_str_has_suffix (path, ".json"))
+    g_debug ("on_machines_changed: event type %i on %s", event_type, path);
+  else
+    {
+      g_debug ("on_machines_changed: ignoring event type %i on non-.json file %s", event_type, path);
+      g_free (path);
+      return;
+    }
+  g_free (path);
+
+  machines = build_machines ();
```

Comment:
This is the callback when `/etc/cockpit/machines.d/` changes, so here we have to rebuild it. Parsing a few small JSON files is rather cheap, I don't want to introduce complex heuristics about which files we don't need to re-parse and keep a per-file partial struct. This would be very involved and error prone.